bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryClientImpl.java

```diff
@@ -0,0 +1,492 @@
+                  dataset)
+              .setRequestHeaders(new HttpHeaders().setIfMatch(dataset.getEtag()))
+              .executeUnparsed();
+      if (response.getStatusCode() == HttpStatusCodes.STATUS_CODE_NOT_FOUND) {
+        throw new NoSuchNamespaceException(response.getStatusMessage());
+      }
+
+      return convertExceptionIfUnsuccessful(response).parseAs(Dataset.class);
+    } catch (IOException e) {
+      throw new RuntimeIOException(e);
+    }
+  }
+
+  /**
+   * Returns true when it is a BigQuery Metastore Iceberg table, defined by having the
+   * ExternalCatalogTableOptions object and a parameter of METADATA_LOCATION_PROP as part of its
+   * parameters map.
+   *
+   * @param table to check
+   */
+  private boolean isValidIcebergTable(Table table) {
+    return table.getExternalCatalogTableOptions() != null
+        && !table.getExternalCatalogTableOptions().isEmpty()
+        && table.getExternalCatalogTableOptions().getParameters() != null
+        && table
+            .getExternalCatalogTableOptions()
+            .getParameters()
+            .containsKey(BaseMetastoreTableOperations.METADATA_LOCATION_PROP)
+        && BaseMetastoreTableOperations.ICEBERG_TABLE_TYPE_VALUE.equalsIgnoreCase(
+            table
+                .getExternalCatalogTableOptions()
+                .getParameters()
+                .get(BaseMetastoreTableOperations.TABLE_TYPE_PROP));
+  }
+
+  private Table validateTable(Table table) {
+    if (!isValidIcebergTable(table)) {
+      throw new NoSuchIcebergTableException("This table is not a valid Iceberg table: %s", table);
+    }
+
+    return table;
+  }
+
+  /**
+   * Converts BigQuery generic API errors to Iceberg exceptions, *without* handling the
+   * resource-specific exceptions like NoSuchTableException, NoSuchNamespaceException, etc.
+   */
+  @SuppressWarnings("FormatStringAnnotation")
+  private HttpResponse convertExceptionIfUnsuccessful(HttpResponse response) throws IOException {
+    if (response.isSuccessStatusCode()) {
+      return response;
+    }
+
+    String errorMessage =
+        String.format(
+            "%s\n%s",
+            response.getStatusMessage(),
+            response.getContent() != null
+                ? new String(response.getContent().readAllBytes(), StandardCharsets.UTF_8)
+                : "");
+    switch (response.getStatusCode()) {
+      case HttpStatusCodes.STATUS_CODE_UNAUTHORIZED:
+        throw new NotAuthorizedException(errorMessage, NOT_AUTHORIZED_ERROR_MESSAGE);
+      case HttpStatusCodes.STATUS_CODE_BAD_REQUEST:
+        throw new BadRequestException(errorMessage);
+      case HttpStatusCodes.STATUS_CODE_FORBIDDEN:
+        throw new ForbiddenException(errorMessage);
+      case HttpStatusCodes.STATUS_CODE_PRECONDITION_FAILED:
+        throw new ValidationException(errorMessage);
+      case HttpStatusCodes.STATUS_CODE_NOT_FOUND:
+        throw new NotFoundException(errorMessage);
```

Comment:
`NotFoundException` is used for files, not for catalog objects. It think this should be `NoSuchTableException` or `NoSuchNamespaceException`. You probably need to provide context from the HTTP call to know which one it should be.

Reply:
Looks like there is separate handling for this. In that case, it should probably not be here. Probably throw `IllegalArgumentException` here because it is invalid to call this method with STATUS_CODE_NOT_FOUND.